### PR TITLE
implement external token access

### PR DESCRIPTION
**Pull request by freben**
https://github.com/backstage/backstage/tree/master/beps/0007-auth-external-services

Ping @mareklibra 

This introduces both the new `static` type tokens that are just plain static strings, as well as the `legacy` type which mimics the old backend-to-backend auth flow.

We have tried to keep cross-compatibility between old and new, by ensuring that the old `ServerTokenManager` is able to always read the same set of secrets as the new code does.

### CODE REVIEW

File: docs/auth/service-to-service-auth.md [hunk 1]
```diff
@@ -2,168 +2,167 @@
...
+
+```yaml title="in e.g. app-config.production.yaml"
 backend:
   auth:
-    keys:
-      - secret: <the string returned by the above crypto command>
-    # - secret: ${BACKEND_SECRET} - if you want to use an env variable instead
+    externalAccess:
+      - type: static
+        config:
```
[the remark this answers is not in the document]
Comment (freben) [reply docs/auth/service-to-service-auth.md]: Good catch! Had a bug that also needed fixing because of this

**Review by Rugvip — CHANGES_REQUESTED** [review]
polished 👌 ✨

File: packages/backend-app-api/src/services/implementations/auth/external/types.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ * limitations under the License.
+ */
+
+import { Config } from '@backstage/config';
+
+export interface TokenHandler {
+  add(options: Config): void;
+  verifyToken(
+    token: string,
+  ): Promise<{ subject: string; token?: string } | undefined>;
```
[the remark this answers is not in the document]
Comment (freben) [reply packages/backend-app-api/src/services/implementations/auth/external/types.ts]: removed

File: packages/backend-app-api/src/services/implementations/auth/external/static.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    if (!token.match(/^\S+$/)) {
+      throw new Error('Illegal token, must be a set of non-space characters');
+    }
+
+    const subject = options.getString('subject');
+    if (!subject.match(/^\S+$/)) {
+      throw new Error('Illegal subject, must be a set of non-space characters');
+    }
+
+    this.#entries.push({ token, subject });
```
[the remark this answers is not in the document]
Comment (freben) [reply packages/backend-app-api/src/services/implementations/auth/external/static.ts:46]: yep, agreed, done

File: packages/backend-app-api/src/services/implementations/auth/DefaultAuthService.ts [hunk 4]
```diff
@@ -78,14 +80,15 @@ export class DefaultAuthService implements AuthService {
...
-    // Legacy service-to-service token
-    const { sub, aud } = decodeJwt(token);
-    if (sub === 'backstage-server' && !aud) {
-      await this.tokenManager.authenticate(token);
-      return createCredentialsWithServicePrincipal('external:backstage-plugin');
+    const externalResult = await this.externalTokenHandler.verifyToken(token);
+    if (externalResult) {
+      return createCredentialsWithServicePrincipal(
+        externalResult.subject,
+        externalResult.token,
```
[the remark this answers is not in the document]
Comment (freben) [reply packages/backend-app-api/src/services/implementations/auth/DefaultAuthService.ts]: removed

File: packages/backend-app-api/config.d.ts [hunk 5]
```diff
@@ -32,6 +32,104 @@ export interface Config {
        * unless you configure credentials for service calls.
        */
       dangerouslyDisableDefaultAuthPolicy?: boolean;
+
+      /**
+       * Configures methods of external access, ie ways for callers outside of
+       * the Backstage ecosystem to get authorized for access to APIs that do
+       * not permit unauthorized access.
+       *
+       * @deepVisibility secret
```
[the remark this answers is not in the document]
Comment (freben) [reply packages/backend-app-api/config.d.ts:41]: Ah that won't do - fixed

**Review by Rugvip — APPROVED** [review]
:shipit:

**Pull request merged**